### PR TITLE
Improve runtime check configurability

### LINKED ISSUE #45: `spec_version` warnings should be logged after migrations are executed
Currently spec version checks are executed and warnings logged before any migrations run, making them easily missed if the migrations are noisy.

The warning should be logged at the end, to reduce the chance of it being missed by the developer.

The process should also exit with a non-zero status code when there is a spec_version warning, unless disabled.

### PR DESCRIPTION
**Pull request by liamaharon**
## Breaking changes
- Rename `--no-idempotency-checks` `--disable-idempotency-checks`
- Changes severity of failing a spec version check from a warning to execution failure (closes https://github.com/paritytech/try-runtime-cli/issues/45)
- Check spec name by default

## Other changes

- Introduce new shared param `check_spec_name`
- Introduce new `OnRuntimeUpgrade` param `check_spec_version`
- Simplify internal handling of runtime checks 

TODO
- [x] Tests

### CODE REVIEW

**Comment by liamaharon** [comment]
@simonsso I can't tag you as a reviewer for some reason, but would appreciate any comment you may have on this pr :)

File: core/src/state.rs [hunk 1]
```diff
@@ -319,25 +305,25 @@ impl State {
                 new_code_hash
             );
 
-            if new_version.spec_name != old_version.spec_name {
+            if runtime_checks.name_matches && new_version.spec_name != old_version.spec_name {
                 return Err("Spec names must match.".into());
+            } else {
+                log::info!("Skipping spec name check.")
```
Comment (bkchr) [inline core/src/state.rs:311]: This is not correct as you may have done the check. If you really want to print this, you need to create an outer if with `runtime_checks.name_matches` only as condition.

File: core/src/state.rs [hunk 2]
```diff
@@ -319,25 +305,25 @@ impl State {
...
+            if runtime_checks.version_increases
                 && new_version.spec_version <= old_version.spec_version
             {
-                log::warn!(
-                    target: LOG_TARGET,
-                    "New runtime spec version is not greater than the on-chain runtime spec version. Don't forget to increment the spec version if you intend to use the new code in a runtime upgrade."
-                );
+                return Err("New runtime spec version must be greater than the on-chain runtime spec version.".into());
+            } else {
+                log::info!("Skipping spec version check.")
```
Comment (bkchr) [inline core/src/state.rs:319]: Same.

File: core/src/commands/on_runtime_upgrade.rs [hunk 3]
```diff
@@ -61,6 +61,10 @@ pub struct Command {
     #[clap(long, default_value = "false", default_missing_value = "true")]
     pub no_weight_warnings: bool,
 
+    /// Whether to enforce the new runtime `spec_version` is greater or equal to the existing
+    /// `spec_version`.
+    #[clap(long, default_value = "true", default_missing_value = "true")]
+    pub check_spec_version: bool,
```
Comment (bkchr) [inline core/src/commands/on_runtime_upgrade.rs:67]:
```suggestion
    #[clap(long)]
    pub disable_check_spec_version: bool,
```

These are flags and you should not pass any `true` or `false`. So make it more clear by naming it `disable-check-spec-version`.

File: core/src/commands/on_runtime_upgrade.rs [hunk 4]
```diff
@@ -61,6 +61,10 @@ pub struct Command {
...
     /// Whether to disable migration idempotency checks
     #[clap(long, default_value = "false", default_missing_value = "true")]
     pub no_idempotency_checks: bool,
```
Comment (bkchr) [inline core/src/commands/on_runtime_upgrade.rs:70]: Same should have been done here.

**Review by bkchr — APPROVED** [review]

**Pull request merged**